File: pythonLeetcode/100-相同的树.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
给定两个二叉树，编写一个函数来检验它们是否相同。

如果两个树在结构上相同，并且节点具有相同的值，则认为它们是相同的。

示例 1:

输入:       1         1
          / \       / \
         2   3     2   3

        [1,2,3],   [1,2,3]

输出: true
示例 2:

输入:      1          1
          /           \
         2             2

        [1,2],     [1,null,2]

输出: false
示例 3:

输入:       1         1
          / \       / \
         2   1     1   2

        [1,2,1],   [1,1,2]

输出: false

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/same-tree
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def sameTwoTree(one,two):
    if not one and two:
        return False
    elif one and not two:
        return False
    elif not one and not two:
        return True
    else:
        if one.val != two.val:
            return False
        elif not sameTwoTree(one.left,two.left):
            return False
        elif not sameTwoTree(one.right,two.right):
            return False
        return True

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        return sameTwoTree(p,q)

File: LoveSwiftLeetcode/100.相同的树.swift
/*
 * @lc app=leetcode.cn id=100 lang=swift
 *
 * [100] 相同的树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isSameTree(_ p: TreeNode?, _ q: TreeNode?) -> Bool {
        if p == nil , q != nil {
            return false
        }
        if p != nil , q == nil {
            return false
        }
        if p == nil , q == nil {
            return true
        }
        if p!.val != q!.val {
            return false
        }
        let lr = isSameTree(p!.left, q!.left)
        if lr == false {
            return false
        }
        let rr = isSameTree(p!.right, q!.right)
        if rr == false {
            return false
        }
        return true
    }
}
// @lc code=end

